### PR TITLE
Add support for specifying a list of autocmd event names to autocmd_add()

### PR DESCRIPTION
**Pull request by yegappan**
The autocmd_add() function now supports either a String with one or more
autocmd event names separated by comma or a List of event names.
It also supports a String or a List for the autocmd pattern.

### CODE REVIEW

File: src/autocmd.c [hunk 1]
```diff
@@ -2794,30 +2802,23 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
-		event = event_name2nr(event_name, &end);
-		if (event == NUM_EVENTS)
-		{
-		    semsg(_(e_no_such_event_str), event_name);
-		    retval = VVAL_FALSE;
-		    break;
-		}
+		event_name = di->di_tv.vval.v_string;
+		if (event_name == NULL)
+		    continue;
```
Comment (LemonBoy) [inline src/autocmd.c:2814]: Empty strings should be an error, no?

File: src/autocmd.c [hunk 2]
```diff
@@ -2794,30 +2802,23 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
+	    }
+	    else if (di->di_tv.v_type == VAR_LIST)
+	    {
+		event_list = di->di_tv.vval.v_list;
+		if (event_list == NULL)
```
Comment (LemonBoy) [inline src/autocmd.c:2820]: Ditto for empty lists (?). And what happens when a range is passed?

File: src/autocmd.c [hunk 3]
```diff
@@ -2794,30 +2802,23 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
+		    continue;
 	    }
+	    else
+		continue;
```
Comment (LemonBoy) [inline src/autocmd.c:2829]: Unsupported type -> error message.

File: src/autocmd.c [hunk 4]
```diff
@@ -2859,21 +2860,31 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
-	    pat = dict_get_string(event_dict, (char_u *)"pattern", TRUE);
-	    if (pat == NULL)
+	    di = dict_find(event_dict, (char_u *)"pattern", -1);
+	    if (di != NULL)
 	    {
-		if (delete)
-		    pat = vim_strsave((char_u *)"");
+		if (di->di_tv.v_type == VAR_STRING)
+		{
+		    pat = di->di_tv.vval.v_string;
```
Comment (LemonBoy) [inline src/autocmd.c:2882]: Ditto as above wrt error conditions.

File: src/autocmd.c [hunk 5]
```diff
@@ -2907,11 +2919,70 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
-		break;
+		if (event_list != NULL)
+		{
+		    if (eli == NULL)
+			eli = event_list->lv_first;
+		    else
+			eli = eli->li_next;
+		    if (eli == NULL)
+			break;
+		    if (eli->li_tv.v_type != VAR_STRING
```
Comment (LemonBoy) [inline src/autocmd.c:2952]: `tv_get_string_strict`? Let's add more checks/errors instead of silently ignoring errors.

File: src/autocmd.c [hunk 6]
```diff
@@ -2907,11 +2919,70 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
+			break;
+		    }
+		}
+		else if (pat_list != NULL)
+		{
+		    FOR_ALL_LIST_ITEMS(pat_list, pli)
+		    {
+			if (pli->li_tv.v_type != VAR_STRING
+				|| pli->li_tv.vval.v_string == NULL)
+			    continue;
```
Comment (LemonBoy) [inline src/autocmd.c:2994]: Ditto, checks checks check.

File: src/autocmd.c [hunk 7]
```diff
@@ -2907,11 +2919,70 @@ autocmd_add_or_delete(typval_T *argvars, typval_T *rettv, int delete)
...
+		{
+		    if (end == NULL)
+			p = end = event_name;
+		    if (end == NULL || *end == NUL)
+			break;
+		}
+		if (p == NULL)
+		    continue;
+
+		event = event_name2nr(p, &end);
```
Comment (LemonBoy) [inline src/autocmd.c:2970]: I don't see any check for trailing characters after the event name, nor checks for trailing commas when passing a string.

**Comment by chrisbra** [comment]
> either a String with one or more autocmd event names separated by comma or a List of event names

Shouldn't we be more consistent here and either allow String for single events or list with multiple events? Or maybe even better disallow strings at all and only allow a list of events (since this is pretty new, shouldn't break users expectations yet I hope)?

**Comment by brammool** [comment]
The most obvious would be allow for a string with a single item, and a list of strings with at least one item.
Each string should be one event name.  I don't think we need to strip white space.

**Pull request closed without merging**